trace-recorder/src/main/org/jetbrains/lincheck/trace/recorder/TraceCollectingEventTracker.kt

```diff
@@ -173,53 +183,97 @@ class TraceCollectingEventTracker(
+
+        fun appendMethodCall(obj: TRObject?, className: String, methodName: String, methodDesc: String, codeLocationId: Int = -1, params: List<TRObject> = emptyList()) {
+            val methodCall = TRIncompleteMethodCallTracePoint(
+                threadId = threadData.threadId,
+                codeLocationId = codeLocationId,
+                methodId = TRACE_CONTEXT.getOrCreateMethodId(className, methodName, methodDesc),
+                obj = obj,
+                parameters = params
+            )
+            val parentTracePoint = if (threadData.getStack().isEmpty()) null
+                                   else threadData.currentMethodCallTracePoint()
+            strategy.tracePointCreated(parentTracePoint, methodCall)
+            threadData.pushStackFrame(methodCall, thread, isInline = false)
+        }
+
+        runInsideInjectedCode {
+            strategy.registerCurrentThread(threadData.threadId)
+            for (frame in thread.stackTrace.reversed()) {
+                if (frame.className == "sun.nio.ch.lincheck.Injections") break
+                if (
+                    !frame.isLincheckInternals &&
+                    !frame.isNativeMethod &&
+                    !analysisProfile.shouldBeHidden(frame.className, frame.methodName)
+                ) {
+                    // TODO: should code locations be computed from the frame?
+                    appendMethodCall(null, frame.className, frame.methodName, "()V")
+                }
+            }
+        }
+    }
+
+    override fun beforeThreadFork(thread: Thread, descriptor: ThreadDescriptor) = runInsideInjectedCode {
         ThreadDescriptor.getCurrentThreadDescriptor() ?: return
         // Create new thread handle
         val forkedThreadData = ThreadData(threads.size)
-        ThreadDescriptor.getThreadDescriptor(thread).eventTrackerData = forkedThreadData
+        val threadDescriptor = ThreadDescriptor.getThreadDescriptor(thread)
+        threadDescriptor.eventTrackerData = forkedThreadData
         threads[thread] = forkedThreadData
         // We are ready to use this
     }
 
     override fun beforeThreadStart() = runInsideIgnoredSection {
         val threadDescriptor = ThreadDescriptor.getCurrentThreadDescriptor() ?: return
         val threadData = threadDescriptor.eventTrackerData as? ThreadData? ?: return
-
-        strategy.registerCurrentThread(threadData.threadId)
-
-        val tracePoint = TRMethodCallTracePoint(
-            threadId = threadData.threadId,
-            codeLocationId = -1,
-            methodId = TRACE_CONTEXT.getOrCreateMethodId("Thread", "run", "()V"),
-            obj = TRObject(Thread.currentThread()),
-            parameters = emptyList()
-        )
-        strategy.tracePointCreated(null, tracePoint)
-        threadData.pushStackFrame(tracePoint, Thread.currentThread(), isInline = false)
+        val thread = Thread.currentThread()
         threadDescriptor.enableAnalysis()
+
+        runInsideInjectedCode {
```

Comment:
Why the whole method is not enclosed in `runInsideInjectedCode` ? 
Could you please add a comment?